Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -111,25 +102,139 @@ FunctionDefinition targetFunction
+            foreach (var lhs in node.Left) {
+                if (lhs is MemberExpression memberExp && memberExp.Target is NameExpression nameExp1) {
+                    if (_classType != null && nameExp1.Name == "self") {
+                        _classType.AddMembers(new[] { new KeyValuePair<string, IMember>(memberExp.Name, value) }, true);
+                    }
+                    continue;
+                }
+
+                if (lhs is NameExpression nameExp2 && nameExp2.Name == "self") {
+                    continue; // Don't assign to 'self'
+                }
+
+                // Basic assignment
+                foreach (var ne in node.Left.OfType<NameExpression>()) {
+                    _scope.SetInScope(ne.Name, value);
+                }
+
+                // Tuple = Tuple. Transfer values.
+                if (lhs is TupleExpression tex) {
+                    if (value is TupleExpression valTex) {
+                        var returnedExpressions = valTex.Items.ToArray();
+                        var names = tex.Items.Select(x => (x as NameExpression)?.Name).ToArray();
+                        for (var i = 0; i < Math.Min(names.Length, returnedExpressions.Length); i++) {
+                            if (returnedExpressions[i] != null) {
+                                var v = _scope.GetValueFromExpression(returnedExpressions[i]);
+                                _scope.SetInScope(names[i], v);
+                            }
+                        }
+                        continue;
+                    }
+
+                    // Tuple = 'tuple value' (such as from callable). Transfer values.
+                    if (value is AstPythonConstant c && c.Type is AstPythonTuple tuple) {
+                        var types = tuple.Types.ToArray();
+                        var names = tex.Items.Select(x => (x as NameExpression)?.Name).ToArray();
+                        for (var i = 0; i < Math.Min(names.Length, types.Length); i++) {
+                            if (names[i] != null) {
+                                _scope.SetInScope(names[i], new AstPythonConstant(types[i]));
+                            }
+                        }
+                    }
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(IfStatement node) {
+            // Handle basic check such as
+            // if isinstance(value, type):
+            //    return value
+            // by assigning type to the value unless clause is raising exception.
+            var ce = node.Tests.FirstOrDefault()?.Test as CallExpression;
+            if (ce?.Target is NameExpression ne && ne?.Name == "isinstance" && ce.Args.Count == 2) {
+                var name = (ce.Args[0].Expression as NameExpression)?.Name;
+                var typeName = (ce.Args[1].Expression as NameExpression)?.Name;
+                if (name != null && typeName != null) {
```

Comment:
We should be able to do a lookup in `_scope` here to get the type name - then this will work for `isinstance(x, TypeDefinedInThisModule)` as well.

Reply:
_scope does not contain builtins so I added extension method  (id from name) next to the name from id for built-ins.